src/include/duckdb/common/types/string.hpp

```diff
@@ -0,0 +1,270 @@
+	static String Copy(const char *data, size_t size) {
+		String result;
+		result.AssignOwning(data, size);
+		return result;
+	}
+
+	static String Copy(const String &other) {
+		return Copy(other.data(), other.size());
+	}
+	static String Copy(const char *data) {
+		return Copy(data, data ? strlen(data) : 0);
+	}
+	static String Copy(const std::string &str) {
+		return Copy(str.data(), str.size());
+	}
+	String Copy() const {
+		return Copy(data(), size());
+	}
+
+	// Creates a new String instance that references the data without owning it
+	// If the size is small enough, it will inline the data; which WILL be owning
+	static String Reference(const char *data, size_t size) {
+		String result;
+		// If we reference, and we can inline it, we make owning anyway
+		if (size <= INLINE_MAX) {
+			result.AssignOwning(data, size);
+		} else {
+			result.ptr = const_cast<char *>(data); // NOLINT allow const cast
+			result.len = size | NON_OWN_BIT;       // Set the non-owning bit
+		}
+		return result;
+	}
+
+	static String Reference(const String &other) {
+		return Reference(other.data(), other.size());
+	}
+	static String Reference(const char *data) {
+		return Reference(data, data ? strlen(data) : 0);
+	}
+	static String Reference(const std::string &str) {
+		return Reference(str.data(), str.size());
+	}
+	String Reference() const {
+		return Reference(data(), size());
+	}
+
+	string ToStdString() const {
+		if (IsInline()) {
+			return string(buf, size());
+		}
+		return string(ptr, size());
+	}
+
+public:
+	static char CharacterToLower(char c) {
+		if (c >= 'A' && c <= 'Z') {
+			return UnsafeNumericCast<char>(c + ('a' - 'A'));
+		}
+		return c;
+	}
+
+	String Lower() const {
+		const auto str_data = data();
+		const auto str_size = size();
+
+		std::string lowercase_str;
+		lowercase_str.reserve(str_size);
+		for (idx_t i = 0; i < str_size; ++i) {
+			lowercase_str.push_back(CharacterToLower(str_data[i]));
+		}
+		return String(lowercase_str);
+	}
+
+private:
+	static constexpr auto INLINE_CAP = sizeof(char *);
+	static constexpr auto INLINE_MAX = INLINE_CAP - 1;
+	static constexpr auto NON_OWN_BIT = 1UL << (sizeof(size_t) * 8 - 1);
+	static constexpr auto LENGTH_MAX = (1UL << (sizeof(size_t) * 8 - 1)) - 1;
+
+	void AssignOwning(const char *new_data, size_t new_size) {
+		len = new_size;
+
+		if (len == 0) {
+			buf[len] = '\0'; // Null-terminate the inline buffer
+			return;
+		}
+
+		D_ASSERT(new_data != nullptr);
+
+		if (len <= INLINE_MAX) {
+			memcpy(buf, new_data, len);
+			buf[len] = '\0'; // Null-terminate the inline buffer
+			return;
+		}
+
+		auto new_ptr = new char[len + 1]; // +1 for null-termination
+		memcpy(new_ptr, new_data, len);
+		new_ptr[len] = '\0';
+
+		ptr = new_ptr;
+	}
+
+	void Destroy() {
+		if (IsOwning() && !IsInline()) {
+			delete[] ptr;
+		}
+	}
+
+	void Release() {
+		// Set the non-owning bit
+		if (IsOwning() && !IsInline()) {
+			len |= NON_OWN_BIT;
+		}
+	}
+
+private:
+	size_t len;
```

Comment:
I think `size_t` is better not used, either `uint32_t` or `uint64_t`? Than it's explicit (and cross platform).

Reply:
I thought size_t was preferable in e.g. wasm. But if we make it explicit I vote for uint32_t. We could use the other 4 bytes in padding to have fun I guess.